### PR TITLE
Additional Optimizations

### PR DESCRIPTION
**Pull request by teubert**
A few more optimizations and improvements, including:
* Updates to Noise to use matrix multiplication inplace (instead of forming a new dictionary)
   * To support this, store process and measurement noise as *Container
* Update to SimResult to store vectorized input as 2d instead of 3d array
   * Add atleast_1d commands in models to support above change (this is only needed in models that support vectorized input- advanced feature)
   * Default next_state uses matrix multiplication if it can
* Extend benchmarking script
* Remove tests with released prog_algs (since that combo won't work)

Note: these changes will speed things up, but will require stricter adherence to using *Containers instead of dictionaries

Note2: This is related to a similar PR in prog_algs https://github.com/nasa/prog_algs/pull/201 (Tests in prog_algs WILL fail before that's merged)

### CODE REVIEW

**Comment by teubert** [comment]
Benchmarking (pre -> post)

import main                             0.9926 -> 1.258566
import thrown object                    0.416 -> 0.5728500000000001
model initialization                    0.0002 -> 0.00036199999999997345
set noise                               2.09e-5 -> 2.9999999999752447e-06
simulate                                0.00044 -> 0.0005100000000000104
simulate with saving                   0.001105 ->  0.0011870000000000491
simulate with saving, dt               0.0024-> 0.00313399999999997
simulate with printing results, dt     0.00366 -> 0.00356600000000018
Plot results                            -> 0.3370899999999999
Metrics                                  ->  0.00039899999999981617
Surrogate Model Generation          ->       0.006158000000000108
surrogate sim                           ->   0.0011570000000000746
surrogate sim, dt                     ->     0.003529999999999589

Speeds up simulation with saving for small dt's (used in prediction), at the cost of slowing imports and initialization (done once)

**Review by kjjarvis — APPROVED** [review]
Edits look good and provide significant improvement in computational efficiency. Just a few comments/questions

File: src/prog_models/utils/noise_functions.py [hunk 1]
```diff
@@ -39,21 +31,13 @@ def no_measurement_noise(self, z : dict) -> dict:
 # ---------------------------
 
 def triangular_process_noise(self, x : dict, dt : int =1):
-    return self.StateContainer({key: x[key] + \
-        dt*np.random.triangular(-self.parameters['process_noise'][key], 0, self.parameters['process_noise'][key], size=None if np.isscalar(x[key]) else len(x[key])) \
-            for key in self.states})
+    return self.StateContainer(x.matrix + np.random.triangular(-1*self.parameters['process_noise'].matrix, 0, self.parameters['process_noise'].matrix, size=x.matrix.shape))
```
[the remark this answers is not in the document]
Comment (teubert) [reply src/prog_models/utils/noise_functions.py]: Good catch- yes that should be there

File: src/prog_models/utils/containers.py [hunk 2]
```diff
@@ -33,7 +33,10 @@ def __reduce__(self):
         return (DictLikeMatrixWrapper, (self._keys, self.matrix))
 
     def __getitem__(self, key : str) -> int:
-        return self.matrix[self._keys.index(key)][0]
+        row = self.matrix[self._keys.index(key)]
+        if len(row) == 1:
+            return self.matrix[self._keys.index(key)][0]
+        return self.matrix[self._keys.index(key)]
```
[the remark this answers is not in the document]
Comment (teubert) [reply src/prog_models/utils/containers.py:39]: Usually we need an else. In this case, we're returning in the if statement, so we do not need the else (the statement after the if will only be executed if the if statement is not)

File: src/prog_models/prognostics_model.py [hunk 3]
```diff
@@ -197,10 +197,11 @@ def initialize(self, u : dict = None, z :dict = None) -> dict:
 
         Example
         -------
-        | m = PrognosticsModel() # Replace with specific model being simulated
-        | u = {'u1': 3.2}
-        | z = {'z1': 2.2}
-        | x = m.initialize(u, z) # Initialize first state
+            :
```
[the remark this answers is not in the document]
Comment (teubert) [reply src/prog_models/prognostics_model.py:200]: This was on purpose. The colon forces the auto documentation generation (sphinx) to create a box for the example- which looks a lot better.

**Pull request merged**